lib/Dialect/TTIR/Transforms/HoistCPUOps.cpp

```diff
@@ -0,0 +1,234 @@
+      operandTypes.push_back(operand.getType());
+    for (auto result : opToHoist->getResultTypes())
+      resultTypes.push_back(result);
+
+    // Create the function signature
+    mlir::FunctionType funcType =
+        mlir::FunctionType::get(context, operandTypes, resultTypes);
+
+    // Create the function in the target module
+    auto hoistedFunc =
+        func::FuncOp::create(opToHoist->getLoc(), functionName, funcType);
+    targetModule.push_back(hoistedFunc);
+
+    // Add a basic block to the function
+    mlir::Block *block = hoistedFunc.addEntryBlock();
+    mlir::OpBuilder builder(block, block->end());
+
+    // Map operands to block arguments and clone the operation
+    llvm::SmallVector<mlir::Value> newOperands;
+    for (auto operand : llvm::enumerate(opToHoist->getOperands())) {
+      newOperands.push_back(block->getArgument(operand.index()));
+    }
+
+    mlir::IRMapping mapping;
+    for (auto operand : llvm::zip(opToHoist->getOperands(), newOperands))
+      mapping.map(std::get<0>(operand), std::get<1>(operand));
+
+    mlir::Operation *clonedOp = builder.clone(*opToHoist, mapping);
+
+    // Add a return operation to the function
+    builder.create<mlir::func::ReturnOp>(opToHoist->getLoc(),
+                                         clonedOp->getResults());
+
+    // Declare the function prototype in the source module
+    localFunc =
+        func::FuncOp::create(opToHoist->getLoc(), functionName, funcType);
+    localFunc.setPrivate(); // Mark as external (no body)
+    sourceModule.push_back(localFunc);
+
+    hoistedFunc->setAttr("arg_ranks", builder.getI64ArrayAttr(ranks));
+  }
+
+  // Replace the original operation with a call to the hoisted function
+  mlir::OpBuilder opBuilder(opToHoist);
+  auto callOp = opBuilder.create<mlir::func::CallOp>(
+      opToHoist->getLoc(), localFunc, opToHoist->getOperands());
+
+  // Replace all results of the original operation with the call results
+  for (auto result : llvm::zip(opToHoist->getResults(), callOp.getResults())) {
+    std::get<0>(result).replaceAllUsesWith(std::get<1>(result));
+  }
+
+  // Erase the original operation
+  opToHoist->erase();
+}
+
+// This pass will collect the operations to be hoisted
+class TTIRHoistAnalyze {
+public:
+  // Custom data structure to store the operations to hoist
+  using HoistOpSet = std::vector<std::set<mlir::Operation *>>;
+
+  // Run the pass and collect operations to hoist
+  TTIRHoistAnalyze(mlir::Operation *op) {
+    auto moduleOp = llvm::dyn_cast<mlir::ModuleOp>(op);
+    assert(moduleOp != nullptr &&
+           "somehow got non-ModuleOp in TTIRHoistAnalyze!");
+
+    // Traverse the operations and look for specific op types
+    // Traverse the module's operations recursively
+    moduleOp.walk([&](mlir::Operation *nestedOp) {
+      if (llvm::isa<ttir::AddOp>(nestedOp) ||
+          llvm::isa<ttir::SubtractOp>(nestedOp) ||
+          llvm::isa<ttir::MultiplyOp>(nestedOp)) {
```

Comment:
This is OK for testing, but we don't want to mainline hoisting all add/sub/mul ops.  For now we should define some pass options that enable explicitly defining which op to hoist based on its location so that we can unit test it.  In the future we can build out the real analysis passes that automatically choose which ones to hoist.

Reply:
Yeah, definitely agree.  I'm not sure I understand what you mean by "based on its location" though, can you clarify that?  I was thinking some sort of hardcoded arguments per instruction type (e.g. a sub-flag for `add` would mark all `ttir.add` ops) kind of thing for testing, but definitely open to other ideas (in practice, that could be useful for testing, but very clunky to scale it to all possible operations etc. so probably not useful in the real world)

Reply:
do you mean actual line number or something like that?